### ISSUE TITLE
10$: Error on redefined generator function in switch

### ISSUE THREAD

**Issue by strager** [quick-lint/quick-lint-js#621]
```js
switch (true) {
  case true:
    function* f() {}
    break;
  case false:
    function* f() {}  // should error (f already declared)
    break;
}
```

**Comment by strager** [comment]
This rule also applies to async functions and async generator functions. It does *not* apply to normal (non-generator non-async) functions.

This rule only applies inside switch statements. It does *not* apply outside switch statements.

**Comment by strager** [comment]
This rule also applies in the following case:
```js
switch (true) {
  case true:
    async function f() {}
    break;
  case false:
    var f;  // should error
    break;
}
```

**Comment by strager** [comment]
Whoever works on this task needs to change this monster conditional: https://github.com/quick-lint/quick-lint-js/blob/aced745dd9e18b5e930a413ac4a6a2ecdad91445/src/lint.cpp#L742-L757

Additionally, because normal functions are treated differently than non-normal for this rule, we need a flag or a new `variable_kind` to distinguish the two.

And, of course, we need to track whether we're in a `switch` statement or not. `linter` doesn't know about `switch` scopes.
